File: src/main/java/com/sjs/lootbotga/game/player/PlayerImpl.java
package com.sjs.lootbotga.game.player;

import com.sjs.lootbotga.game.Table;
import com.sjs.lootbotga.game.cards.Card;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

@Component
public class PlayerImpl implements Player {
	private Integer id;
	private List<Card> hand;
	private List<Card> booty;
	private Map<GameState, Move> strategy;

	public Integer getId() {
		return id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public List<Card> getHand() {
		return hand;
	}

	public void setHand(List<Card> hand) {
		this.hand = hand;
	}

	public List<Card> getBooty() {
		return booty;
	}

	public void setBooty(List<Card> booty) {
		this.booty = booty;
	}

	public Map<GameState, Move> getStrategy() {
		return strategy;
	}

	public void setStrategy(Map<GameState, Move> strategy) {
		this.strategy = strategy;
	}



	public Move planMove(Table table) {
		if (strategy.containsKey(new GameState(getHand(), table.getBattleList(), table.getDeck().isEmpty()))) {
			return strategy.get(new GameState(getHand(), table.getBattleList(), table.getDeck().isEmpty()));
		}
		else {
			Move move = randomValidMove(table);
			strategy.put(new GameState(getHand(), table.getBattleList(), table.getDeck().isEmpty()), move);
			return move;
		}
	}

	private Move randomValidMove(Table table) {
		List<Move> possibleMoves = new ArrayList<Move>();
		for (Card card : hand) {
			possibleMoves.addAll(MoveFactory.findAllPossibleMoves(this, card, table));
		}
		if (possibleMoves.size() > 0) {
			Collections.shuffle(possibleMoves);
			return possibleMoves.get(0);
		}
		else {
			return null;
		}
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		PlayerImpl player = (PlayerImpl) o;

		if (id != null ? !id.equals(player.id) : player.id != null) return false;

		return true;
	}

	@Override
	public int hashCode() {
		return id != null ? id.hashCode() : 0;
	}
}

File: src/main/kotlin/com/sjs/lootbotga/game/player/PlayerImpl.kt
package com.sjs.lootbotga.game.player

import com.sjs.lootbotga.game.Table
import com.sjs.lootbotga.game.cards.Card
import java.util.*

class PlayerImpl : Player{
    private var id: Int? = null
    private var hand: List<Card?>? = null
    private var booty: List<Card?>? = null
    private var strategy: MutableMap<GameState?, Move?>? = null

    override fun getId(): Int? {
        return id
    }

    override fun setId(id: Int?) {
        this.id = id
    }

    override fun getHand(): List<Card?>? {
        return hand
    }

    override fun setHand(hand: List<Card?>?) {
        this.hand = hand
    }

    override override fun getBooty(): List<Card?>? {
        return booty
    }

    override fun setBooty(booty: List<Card?>?) {
        this.booty = booty
    }

    override fun getStrategy(): Map<GameState?, Move?>? {
        return strategy
    }

    fun setStrategy(strategy: MutableMap<GameState?, Move?>?) {
        this.strategy = strategy
    }


    fun planMove(table: Table): Move? {
        return if (strategy!!.containsKey(GameState(getHand(), table.battleList, table.deck.isEmpty()))) {
            strategy!![GameState(getHand(), table.battleList, table.deck.isEmpty())]
        } else {
            val move = randomValidMove(table)
            strategy!![GameState(getHand(), table.battleList, table.deck.isEmpty())] = move
            move
        }
    }

    private fun randomValidMove(table: Table): Move? {
        val possibleMoves: MutableList<Move?> = ArrayList()
        for (card in hand!!) {
            possibleMoves.addAll(MoveFactory.findAllPossibleMoves(this, card, table))
        }
        return if (possibleMoves.size > 0) {
            Collections.shuffle(possibleMoves)
            possibleMoves[0]
        } else {
            null
        }
    }

}